lucille-plugins/lucille-pinecone/src/main/java/com/kmwllc/lucille/pinecone/stage/EmitDeletedByPrefix.java

```diff
@@ -0,0 +1,146 @@
+
+import com.kmwllc.lucille.core.Document;
+import com.kmwllc.lucille.core.Stage;
+import com.kmwllc.lucille.core.StageException;
+import com.kmwllc.lucille.pinecone.util.PineconeManager;
+import com.typesafe.config.Config;
+import io.pinecone.clients.Index;
+import io.pinecone.clients.Pinecone;
+import io.pinecone.proto.ListItem;
+import io.pinecone.proto.ListResponse;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.openapitools.control.client.model.IndexModelStatus.StateEnum;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This stage collects all records from namespace(s) if their prefix ID matches the ID of the Lucille document marked for deletion.
+ * These records are then emitted as Lucille Documents, which are also marked for deletion.
+ *
+ * Config Parameters:
+ * - dropOriginal (Boolean, Optional): if set to true, will drop documents in the pipeline that were marked for deletion
+ *   as dropped. Does not apply to emitted documents, only the original documents. Defaults to false
+ * - addPrefix (String, optional) : String that adds on to all document ids used for listing
+ *    e.g. if id of doc is "doc1" & addPrefix is "-v1", then this stage will retrieve and emit documents where their
+ *         id prefix is "doc1-v1".
+ * - index (String) : name of the index to search within
+ * - apiKey (String) : apiKey to send requests
+ * - namespaces (Map<String, Object>, Optional) : mapping of namespaces of which you would like to collect ids on. If not given, will
+ *    retrieve from default namespace.
+ *    note that value of mapping will not matter, rather makes it convenient for user to reuse same configs as PineconeIndexer
+ * - deletionMarkerField (String) : name of the field which indicates the document is marked for deletion
+ * - deletionMarkerFieldValue (String) : value of the field which indicates that the document is marked for deletion
+ *   - for stage to work, deletionMarkerField and deletionMarkerFieldValue must be the same value provided indexer configs
+ */
+public class EmitDeletedByPrefix extends Stage {
+
+  private final Pinecone client;
+  private final String indexName;
+  private final Index index;
+  private final Map<String, Object> namespaces;
+  private final boolean dropOriginal;
+  private final String addPrefix;
+  private final String deletionMarkerField;
+  private final String deletionMarkerFieldValue;
+
+  private static final Logger log = LoggerFactory.getLogger(EmitDeletedByPrefix.class);
+
+  public EmitDeletedByPrefix(Config config) {
+    super(config, new StageSpec()
+        .withOptionalProperties("dropOriginal", "addPrefix")
+        .withOptionalParents("namespaces")
+        .withRequiredProperties("apiKey", "deletionMarkerField", "deletionMarkerFieldValue", "index"));
+    this.client = PineconeManager.getClient(config.getString("apiKey"));
```

Comment:
The creation of the client should be moved to the start method. Same for anything that relies on calling externally using the client.

In the lifecycle of a stage. The constructor should never fail. The start method should fail if there are issues setting up anything that it depends on e.g. client / index connection and if theres an issue in startup that would prevent it from working throw a stage exception.